homeassistant/components/opower/coordinator.py

```diff
@@ -0,0 +1,205 @@
+"""Coordinator to handle Opower connections."""
+from datetime import datetime, timedelta
+import logging
+from types import MappingProxyType
+from typing import Any, cast
+
+from aiohttp.client_exceptions import ClientResponseError
+from opower import Account, AggregateType, CostRead, Forecast, MeterType, Opower
+
+from homeassistant.components.recorder import get_instance
+from homeassistant.components.recorder.models import StatisticData, StatisticMetaData
+from homeassistant.components.recorder.statistics import (
+    async_add_external_statistics,
+    get_last_statistics,
+    statistics_during_period,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, UnitOfEnergy, UnitOfVolume
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import CONF_UTILITY, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OpowerCoordinator(DataUpdateCoordinator):
+    """Handle fetching Opower data, updating sensors and inserting statistics."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry_data: MappingProxyType[str, Any],
+    ) -> None:
+        """Initialize the data handler."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Opower",
+            update_interval=timedelta(hours=12),
+        )
+        self.api = Opower(
+            aiohttp_client.async_get_clientsession(hass),
+            entry_data[CONF_UTILITY],
+            entry_data[CONF_USERNAME],
+            entry_data[CONF_PASSWORD],
+        )
+
+    async def _async_update_data(
+        self,
+    ) -> dict[str, Forecast]:
+        """Fetch data from API endpoint."""
+        try:
+            await self.api.async_login()
```

Comment:
Do you have to log back in each time you want data? How frequently does the login expire?

Reply:
Yes. Added comment:
            # Login expires after a few minutes.
            # Given the infrequent updating (every 12h)
            # assume previous sessions have expired and re-login.